drivers/counter/counter_mchp_sam_pit64b.c

```diff
@@ -0,0 +1,439 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT microchip_sam_pit64b_counter
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/counter.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/clock_control/atmel_sam_pmc.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(counter_mchp_sam, CONFIG_COUNTER_LOG_LEVEL);
+
+#define PERIOD_MAX UINT64_MAX
+
+struct sam_pit_config {
+	struct counter_config_info info;
+	pit64b_registers_t *regs;
+	const struct atmel_sam_pmc_config clock_cfg;
+	const struct atmel_sam_pmc_config gclk_cfg;
+	uint8_t clock_selection;
+	uint8_t prescaler_period;
+	uint8_t top_alarm;
+	void (*irq_config_func)(const struct device *dev);
+};
+
+struct sam_pit_alarm_data {
+	counter_alarm_callback_t callback;
+	void *user_data;
+};
+
+struct sam_pit_data {
+	counter_top_callback_t top_cb;
+	void *top_user_data;
+	struct k_spinlock lock;
+
+	struct sam_pit_alarm_data alarm;
+};
+
+static inline void pit_configure(pit64b_registers_t *regs, uint32_t mode)
+{
+	regs->PIT64B_MR = mode;
+}
+
+static inline void pit_start(pit64b_registers_t *regs)
+{
+	regs->PIT64B_CR = PIT64B_CR_START_1;
+}
+
+static inline void pit_stop(pit64b_registers_t *regs)
+{
+	regs->PIT64B_CR = PIT64B_CR_SWRST_1;
+}
+
+static inline uint32_t pit_irq_status(pit64b_registers_t *regs)
+{
+	return regs->PIT64B_ISR;
+}
+
+static inline void pit_irq_enable(pit64b_registers_t *regs, uint32_t mask)
+{
+	regs->PIT64B_IER = mask & PIT64B_IER_Msk;
+}
+
+static inline void pit_irq_disable(pit64b_registers_t *regs, uint32_t mask)
+{
+	regs->PIT64B_IDR = mask & PIT64B_IDR_Msk;
+}
+
+static inline void pit_irq_disable_all(pit64b_registers_t *regs)
+{
+	regs->PIT64B_IDR = PIT64B_IDR_Msk;
+}
+
+static inline uint32_t pit_irq_mask(pit64b_registers_t *regs)
+{
+	return regs->PIT64B_IMR;
+}
+
+static inline uint64_t pit_counter_value(pit64b_registers_t *regs)
+{
+	uint64_t ret;
+
+	ret = regs->PIT64B_TLSBR;
+	ret |= (uint64_t)regs->PIT64B_TMSBR << 32;
+
+	return ret;
+}
+
+static inline uint64_t pit_period_get_value(pit64b_registers_t *regs)
+{
+	return (uint64_t)regs->PIT64B_MSBPR << 32 | regs->PIT64B_LSBPR;
+}
+
+static inline void pit_period_set_value(pit64b_registers_t *regs, uint64_t value)
+{
+	regs->PIT64B_MSBPR = (uint32_t)(value >> 32);
+	regs->PIT64B_LSBPR = (uint32_t)value;
+}
+
+static int sam_pit_start(const struct device *dev)
+{
+	const struct sam_pit_config *config = dev->config;
+
+	pit_start(config->regs);
+
+	return 0;
+}
+
+static int sam_pit_stop(const struct device *dev)
+{
+	const struct sam_pit_config *config = dev->config;
+
+	pit_stop(config->regs);
+
+	return 0;
+}
+
+static int sam_pit_get_value(const struct device *dev, uint32_t *ticks)
+{
+	const struct sam_pit_config *config = dev->config;
+	struct sam_pit_data *data = dev->data;
+	k_spinlock_key_t key;
+
+	key = k_spin_lock(&data->lock);
+
+	*ticks = (uint32_t)pit_counter_value(config->regs);
+
+	k_spin_unlock(&data->lock, key);
```

Comment:
Alternative is to use `K_SPINLOCK`, which is a bit cleaner IMO (doesn't work with `return` statements in the block).


```c
	K_SPINLOCK(&data->lock) {
		*ticks = (uint32_t)pit_counter_value(config->regs);
	}
```